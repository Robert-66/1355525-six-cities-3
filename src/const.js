export const MapSettings = {
  ZOOM: 12,
  ICON_URL: `img/pin.svg`,
  ICON_SIZE: [27, 39]
};
