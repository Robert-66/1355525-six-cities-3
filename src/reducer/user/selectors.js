export function getAuthorizationStatus(state) {
  return state.user.authorizationStatus;
}
