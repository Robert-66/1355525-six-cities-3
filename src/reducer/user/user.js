import {extend} from '../../utils';

const AuthorizationStatus = {
  AUTH: `AUTH`,
  NO_AUTH: `NO_AUTH`,
};

const initialState = {
  authorizationStatus: AuthorizationStatus.NO_AUTH,
  email: ``,
};

const ActionType = {
  AUTH: `AUTH`,
  NO_AUTH: `NO_AUTH`,
};

const ActionCreator = {
  auth: (authInfo) => {
    return {
      type: ActionType.AUTH,
      payload: authInfo,
    };
  },
  noAuth: () => {
    return {
      type: ActionType.NO_AUTH,
    };
  },
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.AUTH:
      return extend(state, {authorizationStatus: AuthorizationStatus.AUTH, email: action.payload.email});
    case ActionType.NO_AUTH:
      return initialState;
  }

  return state;
};

const Operation = {
  checkAuth: () => (dispatch, getState, api) => {
    return api.get(`/login`)
      .then((res) => {
        dispatch(ActionCreator.auth(res.data));
      });
  },
  login: (authData) => (dispatch, getState, api) => {
    return api.post(`/login`, {
      email: authData.email,
      password: authData.password,
    })
      .then((res) => {
        dispatch(ActionCreator.auth(res.data));
      });
  },
};


export {
  ActionCreator,
  ActionType,
  AuthorizationStatus,
  Operation,
  reducer,
};
