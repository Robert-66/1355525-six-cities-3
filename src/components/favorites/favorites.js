import React from 'react';
import {connect} from 'react-redux';
import {Link} from 'react-router-dom';
import {getOffersFavoriteGroupsByCity} from '../../reducer/data/offers-favorite/selectors';
import {Operation as OffersFavoriteOperation} from '../../reducer/data/offers-favorite/offers-favorite';
import {Operation as OffersOperation} from '../../reducer/data/offers/offers.js';
import PlaceCardList from '../place-card-list/place-card-list';
import FavoritesEmpty from '../favorites-empty/favorites-empty';
import {PlaceCardView, AppRoute} from '../../const';
import PropTypes from 'prop-types';
import {offerType} from '../../types/offers-types';

class Favorites extends React.PureComponent {
  constructor(props) {
    super(props);
  }

  componentDidMount() {
    this.props.onFetchOffersFavorite();
  }

  render() {
    const {offersGroup, isLoading, isError, onChangeFavoriteStatus} = this.props;

    return (
      <>
        {Object.keys(offersGroup).length > 0 && (
          <main className="page__main page__main--favorites">
            <div className="page__favorites-container container">
              <section className="favorites">
                <h1 className="favorites__title">Saved listing</h1>
                <ul className="favorites__list">
                  {Object.keys(offersGroup).map((city) => (
                    <li key={city} className="favorites__locations-items">
                      <div className="favorites__locations locations locations--current">
                        <div className="locations__item">
                          <a className="locations__item-link" href="#">
                            <span>{city}</span>
                          </a>
                        </div>
                      </div>
                      <PlaceCardList
                        offers={offersGroup[city]}
                        view={PlaceCardView.FAVORITE}
                        onFavoriteClick={onChangeFavoriteStatus}
                      />
                    </li>
                  ))}
                </ul>
              </section>
            </div>
          </main>
        )}
        {Object.keys(offersGroup).length === 0 && !isLoading && !isError && (
          <FavoritesEmpty />
        )}
        <footer className="footer container">
          <Link className="footer__logo-link" to={AppRoute.ROOT}>
            <img className="footer__logo" src="img/logo.svg" alt="6 cities logo" width="64" height="33" />
          </Link>
        </footer>
      </>
    );
  }
}

Favorites.propTypes = {
  offersGroup: PropTypes.shape({
    [PropTypes.string]: PropTypes.arrayOf(offerType.isRequired)
  }).isRequired,
  isLoading: PropTypes.bool.isRequired,
  isError: PropTypes.bool.isRequired,
  onFetchOffersFavorite: PropTypes.func.isRequired,
  onChangeFavoriteStatus: PropTypes.func.isRequired
};

function mapStateToProps(state) {
  return {
    offersGroup: getOffersFavoriteGroupsByCity(state),
    isLoading: state.data.offersFavorite.isLoading,
    isError: state.data.offersFavorite.isError,
  };
}

function mapDispatchToProps(dispatch) {
  return {
    onFetchOffersFavorite: () => dispatch(OffersFavoriteOperation.fetchOffersFavorite()),
    onChangeFavoriteStatus: (offerId, status) => dispatch(OffersOperation.changeOfferFavoriteStatus(offerId, status)),
  };
}

export {Favorites};
export default connect(mapStateToProps, mapDispatchToProps)(Favorites);
