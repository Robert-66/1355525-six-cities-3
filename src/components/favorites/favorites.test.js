import React from 'react';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import {Provider} from 'react-redux';
import {BrowserRouter} from "react-router-dom";
import {Favorites} from './favorites';
import {CityNames} from '../../const';
import {AuthorizationStatus} from "../../reducer/user/user";

const mockStore = configureStore([]);

const offersGroup = {
  [CityNames.AMSTERDAM]: [
    {
      city: {
        name: CityNames.AMSTERDAM,
        location: {
          latitude: 52.37454,
          longitude: 4.897976,
          zoom: 13
        }
      },
      description: `description`,
      bedrooms: 2,
      maxAdults: 3,
      goods: [`Air conditioning`, `Washer`, `Dishwasher`],
      images: [`img/image1.jpg`, `img/image2.jpg`],
      isFavorite: false,
      previewImage: `img/apartment-01.jpg`,
      isPremium: false,
      type: `apartment`,
      id: 11,
      name: `Apartments Prinsengracht`,
      price: 120,
      coords: [52.3909553943508, 4.85309666406198],
      rating: 4.3,
      host: {
        name: `Angelina`,
        isPro: true,
        avatarUrl: `img/avatar-angelina.jpg`,
      },
    },
  ],
  [CityNames.PARIS]: [
    {
      city: {
        name: CityNames.PARIS,
        location: {
          latitude: 48.85661,
          longitude: 2.351499,
          zoom: 13
        }
      },
      description: `description`,
      bedrooms: 1,
      maxAdults: 2,
      goods: [`Air conditioning`, `Washer`, `Dishwasher`],
      images: [`img/image1.jpg`, `img/image2.jpg`],
      isFavorite: true,
      previewImage: `img/apartment-01.jpg`,
      isPremium: true,
      type: `room`,
      id: 51,
      name: `Tile House`,
      price: 125,
      coords: [48.877610000000004, 2.333499],
      rating: 3.5,
      host: {
        name: `Angelina`,
        isPro: true,
        avatarUrl: `img/avatar-angelina.jpg`,
      },
    },
  ],
};

it(`Favorites is rendered correctly`, () => {
  const store = mockStore({
    user: {
      authorizationStatus: AuthorizationStatus.NO_AUTH,
      email: ``,
    }
  });
  const tree = renderer
    .create(
        <Provider store={store}>
          <BrowserRouter>
            <Favorites
              offersGroup={offersGroup}
              isLoading={false}
              isError={false}
              onFetchOffersFavorite={() => {}}
              onChangeFavoriteStatus={() => {}}
            />
          </BrowserRouter>
        </Provider>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
