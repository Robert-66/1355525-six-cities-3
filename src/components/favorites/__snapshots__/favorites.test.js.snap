// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Favorites is rendered correctly 1`] = `
Array [
  <main
    className="page__main page__main--favorites"
  >
    <div
      className="page__favorites-container container"
    >
      <section
        className="favorites"
      >
        <h1
          className="favorites__title"
        >
          Saved listing
        </h1>
        <ul
          className="favorites__list"
        >
          <li
            className="favorites__locations-items"
          >
            <div
              className="favorites__locations locations locations--current"
            >
              <div
                className="locations__item"
              >
                <a
                  className="locations__item-link"
                  href="#"
                >
                  <span>
                    Amsterdam
                  </span>
                </a>
              </div>
            </div>
            <div
              className="favorites__places"
            >
              <article
                className="place-card favorites__card"
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
              >
                <div
                  className="place-card__image-wrapper favorites__image-wrapper"
                  style={
                    Object {
                      "maxWidth": "150px",
                    }
                  }
                >
                  <a
                    href="#"
                  >
                    <img
                      alt="Place image"
                      className="place-card__image"
                      height="200"
                      src="img/apartment-01.jpg"
                      width="260"
                    />
                  </a>
                </div>
                <div
                  className="place-card__info"
                >
                  <div
                    className="place-card__price-wrapper"
                  >
                    <div
                      className="place-card__price"
                    >
                      <b
                        className="place-card__price-value"
                      >
                        €
                        120
                      </b>
                       
                      <span
                        className="place-card__price-text"
                      >
                        / night
                      </span>
                    </div>
                    <button
                      className="button place-card__bookmark-button"
                      onClick={[Function]}
                      type="button"
                    >
                      <svg
                        className="place-card__bookmark-icon"
                        height="19"
                        width="18"
                      >
                        <use
                          xlinkHref="#icon-bookmark"
                        />
                      </svg>
                      <span
                        className="visually-hidden"
                      >
                        To bookmarks
                      </span>
                    </button>
                  </div>
                  <div
                    className="place-card__rating rating"
                  >
                    <div
                      className="place-card__stars rating__stars"
                    >
                      <span
                        style={
                          Object {
                            "width": "86%",
                          }
                        }
                      />
                      <span
                        className="visually-hidden"
                      >
                        Rating
                      </span>
                    </div>
                  </div>
                  <h2
                    className="place-card__name"
                  >
                    <a
                      className="place-card__name-link"
                      href="/offer/11"
                      onClick={[Function]}
                    >
                      Apartments Prinsengracht
                    </a>
                  </h2>
                  <p
                    className="place-card__type"
                  >
                     
                    Apartment
                  </p>
                </div>
              </article>
            </div>
          </li>
          <li
            className="favorites__locations-items"
          >
            <div
              className="favorites__locations locations locations--current"
            >
              <div
                className="locations__item"
              >
                <a
                  className="locations__item-link"
                  href="#"
                >
                  <span>
                    Paris
                  </span>
                </a>
              </div>
            </div>
            <div
              className="favorites__places"
            >
              <article
                className="place-card favorites__card"
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
              >
                <div
                  className="place-card__mark"
                >
                  <span>
                    Premium
                  </span>
                </div>
                <div
                  className="place-card__image-wrapper favorites__image-wrapper"
                  style={
                    Object {
                      "maxWidth": "150px",
                    }
                  }
                >
                  <a
                    href="#"
                  >
                    <img
                      alt="Place image"
                      className="place-card__image"
                      height="200"
                      src="img/apartment-01.jpg"
                      width="260"
                    />
                  </a>
                </div>
                <div
                  className="place-card__info"
                >
                  <div
                    className="place-card__price-wrapper"
                  >
                    <div
                      className="place-card__price"
                    >
                      <b
                        className="place-card__price-value"
                      >
                        €
                        125
                      </b>
                       
                      <span
                        className="place-card__price-text"
                      >
                        / night
                      </span>
                    </div>
                    <button
                      className="button place-card__bookmark-button--active"
                      onClick={[Function]}
                      type="button"
                    >
                      <svg
                        className="place-card__bookmark-icon"
                        height="19"
                        width="18"
                      >
                        <use
                          xlinkHref="#icon-bookmark"
                        />
                      </svg>
                      <span
                        className="visually-hidden"
                      >
                        To bookmarks
                      </span>
                    </button>
                  </div>
                  <div
                    className="place-card__rating rating"
                  >
                    <div
                      className="place-card__stars rating__stars"
                    >
                      <span
                        style={
                          Object {
                            "width": "70%",
                          }
                        }
                      />
                      <span
                        className="visually-hidden"
                      >
                        Rating
                      </span>
                    </div>
                  </div>
                  <h2
                    className="place-card__name"
                  >
                    <a
                      className="place-card__name-link"
                      href="/offer/51"
                      onClick={[Function]}
                    >
                      Tile House
                    </a>
                  </h2>
                  <p
                    className="place-card__type"
                  >
                     
                    Private Room
                  </p>
                </div>
              </article>
            </div>
          </li>
        </ul>
      </section>
    </div>
  </main>,
  <footer
    className="footer container"
  >
    <a
      className="footer__logo-link"
      href="/"
      onClick={[Function]}
    >
      <img
        alt="6 cities logo"
        className="footer__logo"
        height="33"
        src="img/logo.svg"
        width="64"
      />
    </a>
  </footer>,
]
`;
